### PR TITLE
[CBRD-26518] Add SQL test case for ADDDATE/SUBDATE not returning NULL when return_null_on_function_errors=yes  

### PR DESCRIPTION
**Pull request by kimhak-tech**
Refer to http://jira.cubrid.org/browse/CBRD-26518

**Issue:** adddate()/subdate() should return NULL (not ERROR) on function errors when `return_null_on_function_errors=yes`.

There is a related existing test that covers invalid or zero date/time inputs:
https://github.com/CUBRID/cubrid-testcases/blob/develop/sql/_22_news_service_mysql_compatibility/_02_zero_date_time_conversion/cases/zero_date_time_adddate_return_null.sql
This PR will adds additional coverage.

### CODE REVIEW

File: sql/_13_issues/_26_1h/cases/cbrd_26518.sql [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+SELECT DATE_SUB('0001-01-01', INTERVAL 500 DAY);
+
+evaluate 'Case 2.6: NULL propagation in arithmetic expression';
+SELECT ADDDATE('9999-12-31', 1) + 0;
+-- Expected: NULL
+
+evaluate 'Case 2.7: COALESCE wrapping an overflow call';
+SELECT COALESCE(ADDDATE('9999-12-31', 1), DATE'2000-01-01'); 
+-- Expected: '01/01/2000'
+
```
Comment (kwonhoil) [inline sql/_13_issues/_26_1h/cases/cbrd_26518.sql:152]: SELECT GREATEST (ADDDATE('9999-12-31', 1), DATE'2000-01-01');

File: sql/_13_issues/_26_1h/cases/cbrd_26518.sql [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+SELECT *
+FROM (SELECT ADDDATE('9999-12-31', 1) AS result) AS sub
+WHERE sub.result IS NULL;
+-- Expected: NULL
+
+evaluate 'Case 2.11: Multiple overflow calls in the same SELECT';
+SELECT ADDDATE('9999-12-31', 1)  AS col1,
+       SUBDATE('0001-01-01', 500)  AS col2,
+       ADDDATE('0001-01-01', -500) AS col3;
+-- Expected: NULL  NULL  NULL
```
Comment (bagus-kim) [inline sql/_13_issues/_26_1h/cases/cbrd_26518.sql]:
```suggestion
-- Expected: one row, col1 = NULL, col2 = NULL, col3 = NULL
```

**Comment by bagus-kim** [comment]
http://jira.cubrid.org/browse/CBRD-26518 를 보면 하단과 같은 개발자 의견이 존재합니다.
"
SELECT ADDDATE('0001-01-01', -500); 와 같을 때 NULL로 출력되지 않고 에러가 발생함.
또한 ADDDATE에서 계산 된 결과의 오류에 대해 어떤 것은 00-00-0000이 되고 어떤 것은 오류 처리됨

[Case01]
set system parameters 'return_null_on_function_errors=yes';
SELECT ADDDATE('0001-01-01', interval -2 year);
SELECT ADDDATE('0001-01-01', -500);
-> 위 두 SQL문 결과는 '00/00/0000'

SELECT ADDDATE('0001-01-01', interval -1 year);
SELECT ADDDATE('0001-01-01', -2);
-> 위 두 SQL문 결과는 NULL

[Case02]
set system parameters 'return_null_on_function_errors=no';
SELECT ADDDATE('0001-01-01', interval -2 year);
SELECT ADDDATE('0001-01-01', -500);
-> Function called with missing or invalid arguments.

SELECT ADDDATE('0001-01-01', interval -1 year);
SELECT ADDDATE('0001-01-01', -2);
-> 위 두 SQL문 결과는 '00/00/0000'
"

ADDDATE/SUBDATE의 하한 경계 처리에서 zero date 반환과 NULL 반환이 혼재합니다.

매뉴얼에는 계산 결과가 '0000-00-00 00:00:00' ~ '0001-01-01 00:00:00' 사이인 경우
zero DATE/DATETIME 값을 반환한다고 되어 있습니다.

실제 확인 결과:
set system parameters 'return_null_on_function_errors=yes';
- ADDDATE('0001-01-01', INTERVAL -2 YEAR) -> '00/00/0000'
- ADDDATE('0001-01-01', -500) -> '00/00/0000'
- ADDDATE('0001-01-01', INTERVAL -1 YEAR) -> NULL
- ADDDATE('0001-01-01', -2) -> NULL

동일한 하한 underflow 계열에서 일부는 zero date, 일부는 NULL로 처리되어
동작 일관성이 부족해 보입니다.
문서 기준으로는 모두 zero date로 수렴하는 쪽이 더 자연스러워 보이므로,
이 차이가 의도된 스펙인지, 아니면 함수 내부 처리 경로 차이로 인한 결함인지 추가 확인이 필요합니다.

**Comment by bagus-kim** [comment]
-- 상한초과와 하한초과에 대한 시나리오를 추가해봤습니다. 중복되지 않은 시나리오로 작성하려 했는데요. 반영 검토 바랍니다.
-- 그리고 위에서 말씀드린것과 같이 NULL을 기대하고 작성했는데 실제 동작은 zerodate가 나오는 항목이 있습니다. 우선 이것을 확인 후 SQL문은 수정되어야 할 것 같습니다.

set system parameters 'return_null_on_function_errors=yes';

-- ============================================================
-- Case 13: TIMESTAMP boundary behavior
-- 목적:
--   - TIMESTAMP 입력에 대한 ADDDATE/DATE_ADD, SUBDATE/DATE_SUB 동작을 점검
--   - 상한 방향은 DATETIME 결과로 확장되는지 확인
-- 기대:
--   - 상한 초과 방향은 DATETIME 결과로 정상 반환
-- ============================================================

evaluate 'Case 13.1: TIMESTAMP upper boundary by second returns DATETIME';
SELECT
    ADDDATE(timestamp'2038-01-19 03:14:07', INTERVAL 1 SECOND) AS result,
    ADDDATE(timestamp'2038-01-19 03:14:07', INTERVAL 1 SECOND)
        = datetime'2038-01-19 03:14:08.000' AS expect_ok;

evaluate 'Case 13.2: TIMESTAMP upper boundary by day returns DATETIME';
SELECT
    ADDDATE(timestamp'2038-01-19 03:14:07', INTERVAL 1 DAY) AS result,
    ADDDATE(timestamp'2038-01-19 03:14:07', INTERVAL 1 DAY)
        = datetime'2038-01-20 03:14:07.000' AS expect_ok;

evaluate 'Case 13.3: TIMESTAMP alias DATE_ADD also returns DATETIME';
SELECT
    DATE_ADD(timestamp'2038-01-19 03:14:07', INTERVAL 1 SECOND) AS result,
    DATE_ADD(timestamp'2038-01-19 03:14:07', INTERVAL 1 SECOND)
        = datetime'2038-01-19 03:14:08.000' AS expect_ok;

evaluate 'Case 13.4: TIMESTAMP lower safe boundary by second';
SELECT
    SUBDATE(timestamp'1970-01-01 09:00:02', INTERVAL 1 SECOND) AS result,
    SUBDATE(timestamp'1970-01-01 09:00:02', INTERVAL 1 SECOND)
        = datetime'1970-01-01 09:00:01.000' AS expect_ok;

evaluate 'Case 13.5: TIMESTAMP lower safe boundary by zero interval';
SELECT
    ADDDATE(timestamp'1970-01-01 09:00:01', INTERVAL 0 SECOND) AS result,
    ADDDATE(timestamp'1970-01-01 09:00:01', INTERVAL 0 SECOND)
        = datetime'1970-01-01 09:00:01.000' AS expect_ok;

evaluate 'Case 13.6: TIMESTAMP alias DATE_SUB within valid range';
SELECT
    DATE_SUB(timestamp'1970-01-01 09:00:02', INTERVAL 1 SECOND) AS result,
    DATE_SUB(timestamp'1970-01-01 09:00:02', INTERVAL 1 SECOND)
        = datetime'1970-01-01 09:00:01.000' AS expect_ok;


-- ============================================================
-- Case 14: Repeated overflow calls in a single SELECT
-- 목적:
--   - 동일 SELECT 안에서 overflow 함수가 여러 번 호출될 때
--     일부만 NULL 처리되거나 ERROR가 재발하지 않는지 확인
-- 기대:
--   - overflow/underflow 결과는 모두 NULL
-- ============================================================

SELECT ADDDATE('0001-01-01', -1); --'00/00/0000'
SELECT ADDDATE('0001-01-01', -2); --'00/00/0000'
SELECT ADDDATE('0001-01-01', -365); --'00/00/0000'
SELECT ADDDATE('0001-01-01', -366); --'00/00/0000'

SELECT ADDDATE('0001-01-01', INTERVAL -1 DAY); --'00/00/0000'
SELECT ADDDATE('0001-01-01', INTERVAL -2 DAY); --'00/00/0000'
SELECT ADDDATE('0001-01-01', INTERVAL -1 YEAR); --'00/00/0000'
SELECT ADDDATE('0001-01-01', INTERVAL -2 YEAR); --NULL

SELECT SUBDATE('0001-01-01', 1); --'00/00/0000'
SELECT SUBDATE('0001-01-01', 2); --'00/00/0000'
SELECT SUBDATE('0001-01-01', INTERVAL 1 YEAR); --'00/00/0000'
SELECT SUBDATE('0001-01-01', INTERVAL 2 YEAR); --NULL

SELECT DATE_ADD('0001-01-01', INTERVAL -1 YEAR); --'00/00/0000'
SELECT DATE_ADD('0001-01-01', INTERVAL -2 YEAR); --NULL
SELECT DATE_SUB('0001-01-01', INTERVAL 1 YEAR); --'00/00/0000'
SELECT DATE_SUB('0001-01-01', INTERVAL 2 YEAR); --NULL

--


evaluate 'Case 14.1: Same ADDDATE overflow called twice';
SELECT
    ADDDATE('9999-12-31', 1) AS col1,
    ADDDATE('9999-12-31', 1) AS col2,
    ADDDATE('9999-12-31', 1) IS NULL AS col1_null_ok,
    ADDDATE('9999-12-31', 1) IS NULL AS col2_null_ok;

evaluate 'Case 14.2: Same SUBDATE underflow called twice-Expected:NULL?'; 
SELECT
    SUBDATE('0001-01-01', 1) AS col1,
    SUBDATE('0001-01-01', 1) AS col2,
    SUBDATE('0001-01-01', 1) IS NULL AS col1_null_ok, 
    SUBDATE('0001-01-01', 1) IS NULL AS col2_null_ok; 
evaluate 'Case 14.2: Same SUBDATE underflow called twice -Expected:00/00/0000?';
SELECT
    SUBDATE('0001-01-01', 1) AS col1,
    SUBDATE('0001-01-01', 1) AS col2,
    SUBDATE('0001-01-01', 1) = '00/00/0000' AS col1_zero_ok,
    SUBDATE('0001-01-01', 1) = '00/00/0000' AS col2_zero_ok;    

evaluate 'Case 14.3: Mixed repeated overflow and underflow';
SELECT
    ADDDATE('9999-12-31', 1) AS add_over,
    SUBDATE('0001-01-01', 1) AS sub_under,
    ADDDATE('9999-12-31', 1) AS add_over_again,
    SUBDATE('0001-01-01', 1) AS sub_under_again,
    ADDDATE('9999-12-31', 1) IS NULL AS add_over_ok,
    SUBDATE('0001-01-01', 1) IS NULL AS sub_under_ok; 
evaluate 'Case 14.3: Mixed repeated overflow and underflow';
SELECT
    ADDDATE('9999-12-31', 1) AS add_over,
    SUBDATE('0001-01-01', 1) AS sub_under,
    ADDDATE('9999-12-31', 1) AS add_over_again,
    SUBDATE('0001-01-01', 1) AS sub_under_again,
    ADDDATE('9999-12-31', 1) IS NULL AS add_over_ok,
    SUBDATE('0001-01-01', 1) = '00/00/0000' AS sub_under_ok;     

evaluate 'Case 14.4: Repeated alias calls';
SELECT
    DATE_ADD('9999-12-31', INTERVAL 1 DAY) AS col1,
    DATE_SUB('0001-01-01', INTERVAL 1 DAY) AS col2,
    DATE_ADD('9999-12-31', INTERVAL 1 DAY) IS NULL AS col1_null_ok,
    DATE_SUB('0001-01-01', INTERVAL 1 DAY) IS NULL AS col2_null_ok; 
evaluate 'Case 14.4: Repeated alias calls';
SELECT
    DATE_ADD('9999-12-31', INTERVAL 1 DAY) AS col1,
    DATE_SUB('0001-01-01', INTERVAL 1 DAY) AS col2,
    DATE_ADD('9999-12-31', INTERVAL 1 DAY) IS NULL AS col1_null_ok,
    DATE_SUB('0001-01-01', INTERVAL 1 DAY) = '00/00/0000' AS col2_null_ok;

-- ============================================================
-- Case 15: ORDER BY / DISTINCT / GROUP BY context
-- 목적:
--   - 함수 에러가 NULL로 바뀐 뒤 정렬, 중복 제거, 그룹화 과정에서
--     다시 ERROR가 발생하지 않는지 확인
-- 기대:
--   - ERROR 없이 수행
-- ============================================================

evaluate 'Case 15.1: ORDER BY overflow result';
SELECT
    r,
    r IS NULL AS null_ok
FROM (
    SELECT ADDDATE('9999-12-31', 1) AS r
    FROM db_root
) x
ORDER BY r;

-- extected:NULL, result : '00/00/0000'
evaluate 'Case 15.2: ORDER BY underflow result'; 
SELECT
    r,
    r IS NULL AS null_ok
FROM (
    SELECT SUBDATE('0001-01-01', 1) AS r
    FROM db_root
) x
ORDER BY r;

evaluate 'Case 15.3: DISTINCT on overflow result';
SELECT
    r,
    r IS NULL AS null_ok
FROM (
    SELECT DISTINCT ADDDATE('9999-12-31', 1) AS r
    FROM db_root
) x;

evaluate 'Case 15.4: GROUP BY overflow result';
SELECT
    r,
    COUNT(*) AS cnt,
    r IS NULL AS null_ok
FROM (
    SELECT ADDDATE('9999-12-31', 1) AS r
    FROM db_root
) x
GROUP BY r;

evaluate 'Case 15.5: GROUP BY mixed valid and invalid';
SELECT
    x.r,
    COUNT(*) AS cnt,
    CASE WHEN x.r IS NULL THEN 1 ELSE x.r = DATE'9999-12-31' END AS bucket_ok
FROM (
    SELECT ADDDATE('9999-12-31', 1) AS r FROM db_root
    UNION ALL
    SELECT ADDDATE('9999-12-30', 1) AS r FROM db_root
) x
GROUP BY x.r
ORDER BY x.r;


-- ============================================================
-- Case 16: Aggregate context
-- 목적:
--   - overflow 결과 NULL 이 집계 함수와 함께 사용될 때
--     예외가 아닌 정상 NULL/집계 결과로 이어지는지 확인
-- 기대:
--   - ERROR 없이 수행
-- ============================================================

evaluate 'Case 16.1: COUNT on overflow result';
SELECT
    COUNT(ADDDATE('9999-12-31', 1)) AS cnt,
    COUNT(ADDDATE('9999-12-31', 1)) = 0 AS expect_ok
FROM db_root;

evaluate 'Case 16.2: COUNT on underflow result';
SELECT
    COUNT(SUBDATE('0001-01-01', 1)) AS cnt,
    COUNT(SUBDATE('0001-01-01', 1)) = 0 AS expect_ok --NULL기대
FROM db_root;

evaluate 'Case 16.3: MIN and MAX on overflow result';
SELECT
    MIN(ADDDATE('9999-12-31', 1)) AS min_val,
    MAX(ADDDATE('9999-12-31', 1)) AS max_val,
    MIN(ADDDATE('9999-12-31', 1)) IS NULL AS min_null_ok,
    MAX(ADDDATE('9999-12-31', 1)) IS NULL AS max_null_ok
FROM db_root;

evaluate 'Case 16.4: Mixed aggregate valid and invalid';
SELECT
    MIN(x.r) AS min_val,
    MAX(x.r) AS max_val,
    COUNT(x.r) AS cnt_not_null,
    MIN(x.r) = DATE'9999-12-31' AS min_ok,
    MAX(x.r) = DATE'9999-12-31' AS max_ok,
    COUNT(x.r) = 1 AS cnt_ok
FROM (
    SELECT ADDDATE('9999-12-31', 1) AS r FROM db_root
    UNION ALL
    SELECT ADDDATE('9999-12-30', 1) AS r FROM db_root
) x;

evaluate 'Case 16.5: COALESCE with aggregate input';
SELECT
    COUNT(COALESCE(ADDDATE('9999-12-31', 1), DATE'2000-01-01')) AS cnt,
    COUNT(COALESCE(ADDDATE('9999-12-31', 1), DATE'2000-01-01')) = 1 AS expect_ok
FROM db_root;


-- ============================================================
-- Case 17: CASTed input with overflow / underflow
-- 목적:
--   - CAST 이후 함수 평가 단계에서 overflow가 나는 경우에도
--     NULL 처리되는지 확인
-- 기대:
--   - ERROR 없이 NULL 반환
-- 주의:
--   - TIMESTAMP로의 CAST 자체가 유효하지 않은 경우는 본 이슈 범위가 아님
-- ============================================================

evaluate 'Case 17.1: CAST string to DATE then overflow';
SELECT
    ADDDATE(CAST('9999-12-31' AS DATE), 1) AS result,
    ADDDATE(CAST('9999-12-31' AS DATE), 1) IS NULL AS expect_ok;

evaluate 'Case 17.2: CAST string to DATE then underflow';
SELECT
    SUBDATE(CAST('0001-01-01' AS DATE), 1) AS result,
    SUBDATE(CAST('0001-01-01' AS DATE), 1) IS NULL AS expect_ok; --NULL기대

evaluate 'Case 17.3: CAST string to DATETIME then overflow';
SELECT
    ADDDATE(CAST('9999-12-31 23:59:59.999' AS DATETIME), INTERVAL 1 MILLISECOND) AS result,
    ADDDATE(CAST('9999-12-31 23:59:59.999' AS DATETIME), INTERVAL 1 MILLISECOND) IS NULL AS expect_ok;

evaluate 'Case 17.4: CAST string to DATETIME then underflow';
SELECT
    SUBDATE(CAST('0001-01-01 00:00:00.000' AS DATETIME), INTERVAL 1 MILLISECOND) AS result,
    SUBDATE(CAST('0001-01-01 00:00:00.000' AS DATETIME), INTERVAL 1 MILLISECOND) IS NULL AS expect_ok;

evaluate 'Case 17.5: CAST valid TIMESTAMP then safe arithmetic';
SELECT
    ADDDATE(CAST('2038-01-19 03:14:07' AS TIMESTAMP), INTERVAL 1 SECOND) AS result,
    ADDDATE(CAST('2038-01-19 03:14:07' AS TIMESTAMP), INTERVAL 1 SECOND)
        = datetime'2038-01-19 03:14:08.000' AS expect_ok;

evaluate 'Case 17.6: CAST valid TIMESTAMP then safe subtraction'; **
SELECT
    SUBDATE(CAST('1970-01-01 00:00:02' AS TIMESTAMP), INTERVAL 1 SECOND) AS result,
    SUBDATE(CAST('1970-01-01 00:00:02' AS TIMESTAMP), INTERVAL 1 SECOND)
        = datetime'1970-01-01 00:00:01.000' AS expect_ok;


-- ============================================================
-- Case 18: Nested expression stability
-- 목적:
--   - 함수 결과가 NULL일 때 다른 함수/식 안에 중첩되어도
--     다시 ERROR가 발생하지 않는지 확인
-- 기대:
--   - ERROR 없이 NULL 또는 정상 대체값 반환
-- ============================================================

evaluate 'Case 18.1: Nested ADDDATE overflow inside ADDDATE';
SELECT
    ADDDATE(ADDDATE('9999-12-31', 1), 1) AS result,
    ADDDATE(ADDDATE('9999-12-31', 1), 1) IS NULL AS expect_ok;

evaluate 'Case 18.2: Nested SUBDATE underflow inside SUBDATE';
SELECT
    SUBDATE(SUBDATE('0001-01-01', 1), 1) AS result,
    SUBDATE(SUBDATE('0001-01-01', 1), 1) IS NULL AS expect_ok;

evaluate 'Case 18.3: Overflow result inside COALESCE then ADDDATE';
SELECT
    ADDDATE(COALESCE(ADDDATE('9999-12-31', 1), DATE'2000-01-01'), 1) AS result,
    ADDDATE(COALESCE(ADDDATE('9999-12-31', 1), DATE'2000-01-01'), 1)
        = DATE'2000-01-02' AS expect_ok;

evaluate 'Case 18.4: Underflow result inside COALESCE then SUBDATE';
SELECT
    SUBDATE(COALESCE(SUBDATE('0001-01-01', 1), DATE'2000-01-02'), 1) AS result,
    SUBDATE(COALESCE(SUBDATE('0001-01-01', 1), DATE'2000-01-02'), 1)
        = DATE'2000-01-01' AS expect_ok;

evaluate 'Case 18.5: CASE using overflow result as input';
SELECT
    CASE
        WHEN ADDDATE('9999-12-31', 1) IS NULL THEN SUBDATE('2000-01-02', 1)
        ELSE ADDDATE('9999-12-31', 1)
    END AS result,
    CASE
        WHEN ADDDATE('9999-12-31', 1) IS NULL THEN SUBDATE('2000-01-02', 1)
        ELSE ADDDATE('9999-12-31', 1)
    END = DATE'2000-01-01' AS expect_ok;

evaluate 'Case 18.6: Arithmetic expression over COALESCE wrapped overflow';
SELECT
    COALESCE(ADDDATE('9999-12-31', 1), DATE'2000-01-01') + 1 AS result,
    COALESCE(ADDDATE('9999-12-31', 1), DATE'2000-01-01') + 1
        = DATE'2000-01-02' AS expect_ok;


set system parameters 'return_null_on_function_errors=no';
commit;

File: sql/_13_issues/_26_1h/cases/cbrd_26518.sql (hunk 1 above)
Comment (kimhak-tech) [reply sql/_13_issues/_26_1h/cases/cbrd_26518.sql:152]: Reflected!

**Review by kwonhoil — APPROVED** [review]

**Review by ssihil — APPROVED** [review]

**Comment by kimhak-tech** [comment]
@bagus-kim I’ve updated the changes based on your comments. Please feel free to check when you have time.

**Comment by bagus-kim** [comment]
@kimhak-tech 
1) CBRD-26518 에 개발자에게 문의중인 것에 대한 답변이 아직 없습니다.
2) 추가 시나리오 13~18은 추가하지 않나요?

**Review by junsklee — APPROVED** [review]

**Comment by kimhak-tech** [comment]
@bagus-kim 
The additional test cases (13–18) are already included, but the numbering is different from your comment. Here’s the mapping for reference:

> Case 13: TIMESTAMP boundary behavior
> => Case 4.X: TIMESTAMP
> 
> Case 14: DATE lower boundary underflow
> => Case 2.2: DATE lower boundary underflow
> 
> Case 14.X: Repeated overflow calls in a single SELECT
> => Case 8: Repeated overflow calls in a single SELECT
> 
> Case 15: ORDER BY / DISTINCT / GROUP BY context
> => Case 9: ORDER BY / DISTINCT / GROUP BY context
> 
> Case 16: Aggregate context
> => Case 10: Aggregate context
> 
> Case 17: CASTed input with overflow / underflow
> => Case 11: CAST inputs with overflow / underflow
> 
> Case 18: Nested expression stability
> => Case 12: Nested expression stability

I'm sorry for making you confused.

**Review by bagus-kim — APPROVED** [review]

**Review by kiho-um — APPROVED** [review]

**Review by ctshim — APPROVED** [review]

**Comment by kimhak-tech** [comment]
**PS.** For **Case 11.2: CAST string to DATE - underflow**, the CCI result is different because the current CCI implementation does not raise an error. Please refer to the details in [CBRD-25628](http://jira.cubrid.org/browse/CBRD-25628).

Therefore, this test case requires a separate expected answer for CCI.
